optuna/pruners/_patient.py

```diff
@@ -0,0 +1,117 @@
+                clf = SGDClassifier(alpha=alpha)
+                n_train_iter = 100
+
+                for step in range(n_train_iter):
+                    clf.partial_fit(X_train, y_train, classes=classes)
+
+                    intermediate_value = clf.score(X_valid, y_valid)
+                    trial.report(intermediate_value, step)
+
+                    if trial.should_prune():
+                        raise optuna.TrialPruned()
+
+                return clf.score(X_valid, y_valid)
+
+
+            study = optuna.create_study(
+                direction="maximize",
+                pruner=optuna.pruners.PatientPruner(optuna.pruners.MedianPruner(), patience=1),
+            )
+            study.optimize(objective, n_trials=20)
+
+    Args:
+        wrapped_pruner:
+            Wrapped pruner to perform pruning when ``PatientPruner`` allows a
+            trial to be pruned. If it is :obj:`None`, this pruner is equivalent to
+            early-stopping taken the intermediate values in the individual trial.
+        patience:
+            Pruning is disabled until the objective doesn't improve for
+            ``patience`` consecutive steps.
+        min_delta:
+            Tolerance value to check whether or not the objective improves.
+            This value should be non-negative for minimization,
+            and it should be non-positive for maximization.
+    """
+
+    def __init__(
+        self, wrapped_pruner: Optional[BasePruner], patience: int, min_delta: float = 0.0
+    ) -> None:
+
+        if patience < 0:
+            raise ValueError("patience cannot be negative but got {}.".format(patience))
+
+        self.wrapped_pruner = wrapped_pruner
+        self._patience = patience
+        self._min_delta = min_delta
+
+    def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> bool:
+
+        intermediate_values = trial.intermediate_values
+
+        steps = np.asarray(list(intermediate_values.keys()))
+
+        # Do not prune if number of step to determine are insufficient.
+        if steps.size < self._patience + 1:
+            return False
+
+        steps.sort()
+        # This is the score patience steps ago
+        steps_before_patience = steps[: -self._patience]
+        scores_before_patience = np.asarray(
+            list(intermediate_values[step] for step in steps_before_patience)
+        )
+        # And these are the scores after that
+        steps_after_patience = steps[-self._patience :]
+        scores_after_patience = np.asarray(
+            list(intermediate_values[step] for step in steps_after_patience)
+        )
+
+        direction = study.direction
+        if direction == StudyDirection.MINIMIZE:
+            maybe_prune = np.min(scores_before_patience) - self._min_delta < np.min(
+                scores_after_patience
+            )
+        else:
+            maybe_prune = np.max(scores_before_patience) + self._min_delta > np.min(
```

Comment:
Should we take the max here?
```suggestion
            maybe_prune = np.max(scores_before_patience) + self._min_delta > np.max(
```

Reply:
yes! Thank you for suggesting it.